### PR TITLE
hull tests for rancher-cis-benchmark chart

### PR DESCRIPTION
**Pull request by vardhaman22**
linked issues: https://github.com/rancher/rancher/issues/41025

### CODE REVIEW

File: tests/rancher-cis-benchmark/rancher_cis_benchmark.go
Comment (eliyamlevy) [inline tests/rancher-cis-benchmark/rancher_cis_benchmark.go:309]: The only fields that should use MustRenderValue are typically stuff you know cannot be null, like .Chart.Name. Combining two values in one call is also not good. There is a function for values that can be null, it is `checker.RenderValue`.

File: tests/rancher-cis-benchmark/rancher_cis_benchmark.go
Comment (eliyamlevy) [inline tests/rancher-cis-benchmark/rancher_cis_benchmark.go:105]: Please separate cases into individual tests so that there is an increased granularity in the testing coverage. You shouldn't overlap test cases that aren't directly related.

File: tests/rancher-cis-benchmark/rancher_cis_benchmark.go
Comment (eliyamlevy) [reply tests/rancher-cis-benchmark/rancher_cis_benchmark.go:309]: For example this is how I did it.
```
	kubectlRepo, _ := checker.RenderValue[string](tc, ".Values.global.kubectl.repository")
	kubectlTag, _ := checker.RenderValue[string](tc, ".Values.global.kubectl.tag")
	expected := kubectlRepo + ":" + kubectlTag
```

File: tests/rancher-cis-benchmark/rancher_cis_benchmark.go
Comment (eliyamlevy) [inline tests/rancher-cis-benchmark/rancher_cis_benchmark.go]: I would separate this test into 3 or more. For example, one for the EnvVars, another for the CIS Image fields, another for affinity, and last one for resources.

My reasoning is that it's easier to miss coverage when everything is lumped together. We also want the tests to be readable and easy to add to.

File: tests/common/global_psp_checks.go
Comment (eliyamlevy) [inline tests/common/global_psp_checks.go]: This test could be done a bit neater and more readable.

```
Checks: test.Checks{
				checker.PerResource[*rbacv1.ClusterRole](func(tc *checker.TestContext, cr *rbacv1.ClusterRole) {
		pspsEnabled, _ := checker.RenderValue[bool](tc, ".Values.global.cattle.psp.enabled")
		pspsFound := false
		for _, rule := range cr.Rules {
			for _, resource := range rule.Resources {
				if resource == "podsecuritypolicies" {
					pspsFound = true
				}
			}
		}
		if pspsEnabled {
			assert.True(tc.T, pspsFound, "ClusterRole %s has incorrect PSP configuration", cr.Name)
		} else {
			assert.False(tc.T, pspsFound, "ClusterRole %s has incorrect PSP configuration", cr.Name)
		}
	}),
	checker.OnResources[*policyv1beta1.PodSecurityPolicy](func(tc *checker.TestContext, psps []*policyv1beta1.PodSecurityPolicy) {
		pspsEnabled, _ := checker.RenderValue[bool](tc, ".Values.global.cattle.psp.enabled")
		if pspsEnabled {
			assert.Equal(tc.T, 2, len(psps), "Missing PSPs")
		} else {
			assert.Equal(tc.T, 0, len(psps), "Missing PSPs")
		}
	}),
},
```

File: tests/common/helpers.go
Comment (eliyamlevy) [inline tests/common/helpers.go]: I'd switch to RenderValue since this can be an empty string.

File: tests/common/workload_checks.go
Comment (eliyamlevy) [inline tests/common/workload_checks.go]: Why not switch this to `PerWorkloads`?

File: tests/rancher-cis-benchmark/rancher_cis_benchmark.go
Comment (vardhaman22) [reply tests/rancher-cis-benchmark/rancher_cis_benchmark.go:309]: done

File: tests/rancher-cis-benchmark/rancher_cis_benchmark.go
Comment (vardhaman22) [reply tests/rancher-cis-benchmark/rancher_cis_benchmark.go:105]: done

File: tests/rancher-cis-benchmark/rancher_cis_benchmark.go
Comment (vardhaman22) [reply tests/rancher-cis-benchmark/rancher_cis_benchmark.go]: done

File: tests/common/global_psp_checks.go
Comment (vardhaman22) [reply tests/common/global_psp_checks.go]: done, refactored it.

File: tests/common/helpers.go
Comment (vardhaman22) [reply tests/common/helpers.go]: done

File: tests/common/workload_checks.go
Comment (vardhaman22) [reply tests/common/workload_checks.go]: changed to PerWorkloads

**Review by eliyamlevy — COMMENTED** [review]
I think we only need to add one pr which has these "common" test cases. Please sync with @rayandas on this.

https://github.com/rancher/charts/pull/2512#event-8894018224

File: tests/rancher-cis-benchmark/rancher_cis_benchmark.go
Comment (eliyamlevy) [inline tests/rancher-cis-benchmark/rancher_cis_benchmark.go]: This can become one .Set instead of four SetValues
```
Set("image", map[string]string {
    "securityScan.repository": "test/security-scan",
    ...
}),
```

File: tests/rancher-cis-benchmark/rancher_cis_benchmark.go
Comment (eliyamlevy) [inline tests/rancher-cis-benchmark/rancher_cis_benchmark.go]: This can be combined into one `.Set()` call

Comment (eliyamlevy) [reply tests/rancher-cis-benchmark/rancher_cis_benchmark.go]: Happens in a few more places.

**Comment by rayandas** [comment]
@eliyamlevy yeah I can do a rebase of my PR when this will be merged or I can remove the common test cases from my PR

**Comment by eliyamlevy** [comment]
Sounds good, and when you get the time theres still two comments that need to be addressed.

File: tests/rancher-cis-benchmark/rancher_cis_benchmark.go
Comment (vardhaman22) [reply tests/rancher-cis-benchmark/rancher_cis_benchmark.go]: done

File: tests/rancher-cis-benchmark/rancher_cis_benchmark.go
Comment (vardhaman22) [reply tests/rancher-cis-benchmark/rancher_cis_benchmark.go]: fixed in all places.

**Review by eliyamlevy — DISMISSED** [review]
LGTM, you might want to get a second sign off from @aiyengar2 before merging

File: tests/common/global_psp_checks.go
Comment (aiyengar2) [inline tests/common/global_psp_checks.go]: PSP should be capitalized?

Comment (aiyengar2) [reply tests/common/global_psp_checks.go]: Can we rename this check? I'd expect it to be something like`EnsurePSPsExist(numExpectedPSPs int)`

File: tests/common/global_psp_checks.go
Comment (aiyengar2) [inline tests/common/global_psp_checks.go]: nit: error message formatting

File: tests/common/global_psp_checks.go
Comment (aiyengar2) [inline tests/common/global_psp_checks.go]: To avoid duplicating logic, perhaps it might make sense to introduce an RBAC-related looper like we do for workloads here: https://github.com/rancher/hull/blob/824596116fec7cf909442e4186d210fe9022a1e3/pkg/checker/loop.go#L37-L70

cc: @eliyamlevy

**Review by aiyengar2 — CHANGES_REQUESTED** [review]
haven't done a full review yet, but here are my preliminary comments

**Review by eliyamlevy — APPROVED** [review]
LGTM For an initial commit, Still need to figure out how to integrate this into the CI

**Review by aiyengar2 — APPROVED** [review]
didn’t get a chance to fully review this, but approving since it’s already been approved to unblock merge

**Pull request merged**
